### DISCUSSION TITLE
Reading a line of text from file

### DISCUSSION THREAD

**Discussion by dnbsd** [deepnoodle-ai/risor#252] [Q&A] [answered]
I wrote a script mimicking the behavior of interactive programs by reading from stdin. The script:

```go
for {
	printf("> ")
	line := string(os.stdin.read())

	switch line {
	case "generate":
		print("generate")
	case "exit":
		os.exit(0)
	default:
		printf("unknown command: %q\n", line)
	}
}
```

This has a major drawback of not recognizing the newline as the "stop" character. I wonder whether I am missing some information that I glazed over reading the documentation. In Go this would be implemented as such (taken from SO):

```go
reader := bufio.NewReader(os.Stdin)
fmt.Print("Enter text: ")
text, _ := reader.ReadString('\n')
fmt.Println(text)
```

**Comment by mna** [comment, marked as answer]
It seems that `os.stdin.read()` without argument will read until the EOF, so it stops only after ctrl-D. This program seems to do what you want, passing a byte slice to `read` so that it fills it until a newline is found:

```
buf := byte_slice(10)

line := ""
for {
	printf("> ")

	line += string(os.stdin.read(buf))
	for {
		// interestingly, for !strings.contains(line, "\n") {...} does not work
		if strings.contains(line, "\n") {
			break
		} else {
			line += string(os.stdin.read(buf))
		}
	}
	parts := strings.split(line, "\n")
	cmd := switch len(parts) {
		case 0:
			line = ""
			""
		case 1:
			line = ""
			parts[0]
		case 2:
			line = parts[1]
			parts[0]
	}

	switch cmd {
	case "echo":
		print("echo")
	case "exit":
		os.exit(0)
	default:
		printf("unknown command: %q\n", cmd)
	}
}
```

**Reply by dnbsd** [reply]
Thank you, @mna. Your example does indeed solve my problem. Here's how I adapted the gist of your code to a function:

```go
func readLine(f) {
	buf := byte_slice(255)
	for {
		printf("> ")
		line := string(f.read(buf))
		if !strings.contains(line, '\n') {
			continue
		}
		return line[:-1]
	}
}
```

What was the indication that `read` accepts a buffer? Even now, knowing that it exists, I can't find this mentioned anywhere in the documentation.

**Reply by mna** [reply]
> What was the indication that `read` accepts a buffer? Even now, knowing that it exists, I can't find this mentioned anywhere in the documentation.

Yeah it doesn't seem like the `File` object is documented at the moment. I looked at the code and you can see in the `File`'s attributes that `read` has various overloads: https://github.com/risor-io/risor/blob/b83ab51d638a6041e7e0e10e002b69f1e1c8e080/object/file.go#L64-L106.
